File: samples/Linux/netcore_for_Linux_x64_only/netcore1.0/TableManagerSamples/If-then/src/TableListener.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text;
using System.Threading;
using fxcore2;

namespace IfThen
{
    class TableListener : IO2GTableListener
    {
        private ResponseListener mResponseListener;
        private List<string> mRequestIDs;
        private string mOrderID;

        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="responseListener"></param>
        public TableListener(ResponseListener responseListener)
        {
            mResponseListener = responseListener;
            mRequestIDs = new List<string>();
            mOrderID = string.Empty;
        }

        public void SetRequestIDs(List<string> requestIDs)
        {
            mRequestIDs.Clear();
            foreach (string sOrderID in requestIDs)
            {
                mRequestIDs.Add(sOrderID);
            }
        }

        public string GetOrderID()
        {
            return mOrderID;
        }

        #region IO2GTableListener Members

        // Implementation of IO2GTableListener interface public method onAdded
        public void onAdded(string sRowID, O2GRow rowData)
        {
            if (rowData.TableType == O2GTableType.Orders)
            {
                O2GOrderRow orderRow = (O2GOrderRow)rowData;
                if (mRequestIDs.Contains(orderRow.RequestID))
                {
                    Console.WriteLine("The order has been added. OrderID={0}, Type={1}, Rate={2}", orderRow.OrderID, orderRow.Type, orderRow.Rate);
                    mRequestIDs.Remove(orderRow.RequestID);
                }
                if (mRequestIDs.Count == 0)
                {
                    mResponseListener.StopWaiting();
                }
            }
        }

        // Implementation of IO2GTableListener interface public method onChanged
        public void onChanged(string sRowID, O2GRow rowData)
        {
        }

        // Implementation of IO2GTableListener interface public method onDeleted
        public void onDeleted(string sRowID, O2GRow rowData)
        {
        }

        public void onStatusChanged(O2GTableStatus status)
        {
        }

        #endregion

        public void SubscribeEvents(O2GTableManager manager)
        {
            O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.Orders);
            ordersTable.subscribeUpdate(O2GTableUpdateType.Insert, this);
        }

        public void UnsubscribeEvents(O2GTableManager manager)
        {
            O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.Orders);
            ordersTable.unsubscribeUpdate(O2GTableUpdateType.Insert, this);
        }

    }
}

File: samples/Windows/cpp/TableManagerSamples/CreateOCO/source/TableListener.cpp
#include "stdafx.h"
#include <algorithm>
#include "ResponseListener.h"
#include "TableListener.h"

TableListener::TableListener(ResponseListener *responseListener)
{
    mRefCount = 1;
    mResponseListener = responseListener;
    mResponseListener->addRef();
    std::cout.precision(2);
}

TableListener::~TableListener(void)
{
    if (mResponseListener)
        mResponseListener->release();
}

long TableListener::addRef()
{
    return InterlockedIncrement(&mRefCount);
}

long TableListener::release()
{
    InterlockedDecrement(&mRefCount);
    if(mRefCount == 0)
        delete this;
    return mRefCount;
}

/** Set request. */
void TableListener::setRequestIDs(std::vector<std::string> &requestIDs)
{
    mRequestIDs.resize(requestIDs.size());
    std::copy(requestIDs.begin(), requestIDs.end(), mRequestIDs.begin());
}

void TableListener::onAdded(const char *rowID, IO2GRow *row)
{
    if (row->getTableType() == Orders)
    {
        IO2GOrderRow *order = static_cast<IO2GOrderRow *>(row);
        std::vector<std::string>::iterator iter;
        iter = std::find(mRequestIDs.begin(), mRequestIDs.end(), order->getRequestID());
        if (iter != mRequestIDs.end())
        {
            std::cout << "The order has been added. OrderID='" << order->getOrderID() << "', "
                    << "Type='" << order->getType() << "', "
                    << "BuySell='" << order->getBuySell() << "', "
                    << "Rate='" << order->getRate() << "', "
                    << "TimeInForce='" << order->getTimeInForce() << "'"
                    << std::endl;
            mRequestIDs.erase(iter);
            if (mRequestIDs.size() == 0)
                mResponseListener->stopWaiting();
        }
    }
}

void TableListener::onChanged(const char *rowID, IO2GRow *row)
{
}

void TableListener::onDeleted(const char *rowID, IO2GRow *row)
{
}

void TableListener::onStatusChanged(O2GTableStatus status)
{
}

void TableListener::subscribeEvents(IO2GTableManager *manager)
{
    O2G2Ptr<IO2GOrdersTable> ordersTable = (IO2GOrdersTable *)manager->getTable(Orders);

    ordersTable->subscribeUpdate(Insert, this);
}

void TableListener::unsubscribeEvents(IO2GTableManager *manager)
{
    O2G2Ptr<IO2GOrdersTable> ordersTable = (IO2GOrdersTable *)manager->getTable(Orders);

    ordersTable->unsubscribeUpdate(Insert, this);
}
